### ISSUE TITLE
Tokens lost when token or token manager don't exist at destination

### ISSUE THREAD

**Issue by SKYBITDev3** [axelarnetwork/interchain-token-service#103]
I tried an `interchainTransfer` for these cases:
- Token contract exists at destination, but counterpart token manager (that has same address as in source) doesn't exist;
- Token contract nor token manager exists at destination.

In both cases `interchainTransfer` operated normally on the source side, burning tokens.
But, of course, nothing appears on the destination.
So the tokens have basically been lost!

Please add checks for token contract and token manager at destination, and only proceed with execution of `interchainTransfer` if they exist.

**Comment by re1ro** [comment]
Tokens are not lost. Even if someone sends tokens where manager is not deployed (it's possible in the UI) they are just going to be pending as a GMP call approval. So that user can deploy the manager as its pemissionless and then claim the tokens anyways.

**Comment by SKYBITDev3** [comment]
When the token manager doesn't exist at the specified address (i.e. wrong address passed into `setTokenManager`) I only see an Error, nothing pending in Axelarscan:
![image](https://github.com/axelarnetwork/interchain-token-service/assets/141273631/6ed2264e-cb2c-48e1-8706-04bf6902a00e)


Here's a comparison of internal transactions with a successful interchain transfer:

![image](https://github.com/axelarnetwork/interchain-token-service/assets/141273631/8805d0b6-7a90-4f4d-8377-050bab3b1fc5)

When it tries to access the non-existent token manager it reverts. Are tokens lost?

**Comment by re1ro** [comment]
Nope. 
The GMP call approval is still pending at the destination chain Gateway and can be retried again when Token Manager is deployed. Axelar scan just shows that initial attempt to execute was reverted, that means that GMP call approval was not consumed and can be retried again.

**Comment by SKYBITDev3** [comment]
> can be retried

Is the Execute button that appears under the last milestone icon on Axelarscan for retrying?

I updated the token manager address in the token to the correct one and pressed the execute button. But I then get the error message:
> Transaction execution was reverted. Please check the implementation of the destination contract's execute function.

![image](https://github.com/axelarnetwork/interchain-token-service/assets/141273631/1dba9326-612b-4201-86ff-1a8e578e179d)

**Comment by re1ro** [comment]
I think the problem there could be with the way you implement your custom token and integrate it with the ITS.

Try following to see if the issue is related to the ITS itself:
* `deployAndRegisterStandardizedToken` and then send it to another chain. 
*  It should fail at the destination.
* `deployAndRegisterRemoteStandardizedToken` and wait for it to complete on destination.
* Then retry to execute the token transfer again at the destination.
* If it still fails, please share the Axelarscan links to these transactions and we will investigate it on our side.

**Comment by SKYBITDev3** [comment]
I tried retry with a StandardizedToken and it worked.
I then tried retry with a new custom token and it worked.
I then investigated the failed interchain transfer from a month ago and found that I needed to deploy a token manager from the token contract itself (because that's what was done on the source blockchain). Retry worked after that.

I found that deploying the token manager by calling `deployCustomTokenManager(salt)` from a hardhat script results in a different token manager address even with the same `salt`, because it passes `msg.sender` to `getCustomTokenId` in 
https://github.com/axelarnetwork/interchain-token-service/blob/8d66c0b4d42fa7c3918ee9e2ccb9d80b7a4e3b1b/contracts/interchain-token-service/InterchainTokenService.sol#L341-L342. `msg.sender` is the token's contract address if `deployCustomTokenManager` is called from the token, or the user's account address if called from a hardhat script.

So lessons learned is that for interchain transfer on ITS to work, the deployment methods of the token and its token manager must be the same across the different blockchains. The same account must also be used across the different blockchains, so for future-proofing, the private key must be preserved and kept safe for deployments of the token to new blockchains in future.
